pkg/slicegwrecycler/recycler.go

```diff
@@ -0,0 +1,207 @@
+			}
+		}
+	}
+	sliceGwNodePorts, _ = getNodePorts(ctx, meshClient, sliceGw)
+	err := hubClient.UpdateNodePortForSliceGwServer(ctx, sliceGwNodePorts, sliceGwName)
+	if err != nil {
+		log.Error(err, "Failed to update NodePort for sliceGw in the hub")
+		utils.RecordEvent(ctx, eventRecorder, sliceGw, nil, ossEvents.EventSliceGWNodePortUpdateFailed, controllerName)
+		return true, ctrl.Result{}, sliceGwNodePorts, err
+	}
+	return false, reconcile.Result{}, sliceGwNodePorts, nil
+}
+
+func createHeadlessServiceForGwServer(slicegateway *kubeslicev1beta1.SliceGateway) *corev1.Service {
+	svc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      slicegateway.Status.Config.SliceGatewayRemoteGatewayID,
+			Namespace: slicegateway.Namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports:     nil,
+			Selector:  nil,
+			ClusterIP: "None",
+		},
+	}
+	// ctrl.SetControllerReference(slicegateway, svc, r.Scheme)
+	return svc
+}
+
+// getRemoteGwPodName returns the remote gw PodName.
+func getRemoteGwPodName(ctx context.Context, meshClient client.Client, gwRemoteVpnIP string, podIP string) (string, error) {
+	log := logger.FromContext(ctx)
+	log.Info("calling gw sidecar to get PodName", "type", "slicegw")
+	sidecarGrpcAddress := podIP + ":5000"
+	workerGWClient, err := sidecar.NewWorkerGWSidecarClientProvider()
+	if err != nil {
+		os.Exit(1)
+	}
+	remoteGwPodName, err := workerGWClient.GetSliceGwRemotePodName(ctx, gwRemoteVpnIP, sidecarGrpcAddress)
+	log.Info("slicegw remote pod name", "slicegw", remoteGwPodName)
+	if err != nil {
+		log.Error(err, "Failed to get slicegw remote pod name. PodIp: %v", podIP)
+		return "", err
+	}
+	return remoteGwPodName, nil
+}
+
+func getNumberOfGatewayNodePortServices(ctx context.Context, meshClient client.Client, sliceGw *kubeslicev1beta1.SliceGateway) (int, error) {
+	listOpts := []client.ListOption{
+		client.MatchingLabels(map[string]string{
+			controllers.ApplicationNamespaceSelectorLabelKey: sliceGw.Spec.SliceName,
+			"kubeslice.io/slicegw":                           sliceGw.Name,
+		}),
+		client.InNamespace(controllers.ControlPlaneNamespace),
+	}
+	services := corev1.ServiceList{}
+	if err := meshClient.List(ctx, &services, listOpts...); err != nil {
+		return 0, err
+	}
+	return len(services.Items), nil
+}
+
+func serviceForGateway(g *kubeslicev1beta1.SliceGateway, i int) *corev1.Service {
+	svc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "svc-" + g.Name + "-" + fmt.Sprint(i),
+			Namespace: g.Namespace,
+			Labels: map[string]string{
+				controllers.ApplicationNamespaceSelectorLabelKey: g.Spec.SliceName,
+				"kubeslice.io/slicegw":                           g.Name,
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Type:     "NodePort",
+			Selector: labelsForSliceGwService(g.Name, i),
+			Ports: []corev1.ServicePort{{
+				Port:       11194,
+				Protocol:   corev1.ProtocolUDP,
+				TargetPort: intstr.FromInt(11194),
+			}},
+		},
+	}
+	//TODO:rework here
```

Comment:
What rework is planned here?

Reply:
This comment was already there in FSM code, i just copy pasted here same.